cmd/scip/test.go

```diff
@@ -0,0 +1,369 @@
+		},
+		Action: func(c *cli.Context) error {
+			dir := c.Args().Get(0)
+			return testMain(dir, testFlags)
+		},
+	}
+	return test
+}
+
+func testMain(directory string, flags testFlags) error {
+	index, err := readFromOption(flags.from)
+	if err != nil {
+		return err
+	}
+
+	hasFailure := false
+
+	for _, document := range index.Documents {
+		sourceFilePath := filepath.Join(directory, document.RelativePath)
+
+		data, err := os.ReadFile(sourceFilePath)
+		if err != nil {
+			return err
+		}
+
+		failures := []string{}
+		successCount := 0
+
+		lines := strings.Split(string(data), "\n")
+		for lineNumber := 0; lineNumber < len(lines); lineNumber++ {
+			testCasesAtLine, usedLines := testCasesForLine(lineNumber, lines, flags.commentSyntax)
+
+			// if the test file contains no test lines, skip it. Only test the lines
+			// that the test file dictates should be tested
+			if len(testCasesAtLine) == 0 {
+				continue
+			}
+
+			attributes := attributesForOccurrencesAtLine(lineNumber, document.Occurrences)
+			for _, testCase := range testCasesAtLine {
+				if !isValidTestCase(testCase, attributes) {
+					failures = append(failures, formatFailure(lineNumber, testCase, attributes))
+				} else {
+					successCount++
+				}
+			}
+
+			lineNumber += usedLines
+		}
+
+		if len(failures) > 0 {
+			hasFailure = true
+			fmt.Printf(red("✗ %s\n"), document.RelativePath)
+			for _, failure := range failures {
+				fmt.Println(indent(failure, 4))
+			}
+		} else {
+			fmt.Printf(green("✓ %s (%d assertions)\n"), document.RelativePath, successCount)
+		}
+	}
+
+	if hasFailure {
+		return cli.Exit("", 1)
+	}
+
+	return nil
+}
+
+// symbolAttribute refers to a single attribute of a symbol.
+// This can be a definition, reference, documentation, or diagnostic
+type symbolAttribute struct {
+	// the column number where this symbol starts
+	start int
+
+	// the length of the symbol's name
+	length int
+
+	// the type of attribute that this is
+	kind string
+
+	// contextual information about the attribute, as determined
+	// by the [kind]
+	data string
+
+	// any additional information necessary to represent this attribute
+	// each line should be considered a "newline", and is used for multiline
+	// comments (in documentation), and diagnostic information
+	additionalData []string
+}
+
+// symbolAttributeTestCase refers to metadata used to validate
+// [symbolAttributes]
+type symbolAttributeTestCase struct {
+	attribute     *symbolAttribute
+	enforceLength bool
+}
+
+// commentsForLine returns the list of lines, after a provided [lineNumber], which are
+// classified as comment. The comment type can be configured using [commentSyntax].
+//
+// Returns the list of symbolAttributeTestCase(s) for the provided line, and the number of
+// of lines that were "consumed" by the cases on this line
+func testCasesForLine(lineNumber int, lines []string, commentSyntax string) ([]*symbolAttributeTestCase, int) {
+	testCases := []*symbolAttributeTestCase{}
```

Comment:
AFAICT there is no need to use pointers here since the `nil` case isn't possible so it'd be simpler to remove the extra `*` and `&`